napari/components/_layer_slicer.py

```diff
@@ -0,0 +1,103 @@
+import logging
+import weakref
+from concurrent.futures import Executor, Future, ThreadPoolExecutor
+from threading import Lock
+from typing import Dict, Iterable, Tuple
+
+from napari.components import Dims
+from napari.layers import Layer
+from napari.utils.events.event import EmitterGroup, Event
+
+LOGGER = logging.getLogger("napari.components._layer_slicer")
+
+
+class _LayerSlicer:
+    def __init__(self):
+        self.events = EmitterGroup(source=self, ready=Event)
+        self._executor: Executor = ThreadPoolExecutor(max_workers=1)
+        self._layers_to_task: Dict[Tuple[Layer], Future] = {}
+        self.lock = Lock()
+
+    def slice_layers_async(
+        self, layers: Iterable[Layer], dims: Dims
+    ) -> Future[dict]:
+        """This should only be called from the main thread.
+
+        Creates a new task and adds it to the _layers_to_task dict. Cancels
+        all tasks currently pending for that layer.
+        """
+        # lock to protect against mutations of `task_to_layer` in `_on_slice_done`
+        with self.lock:
+            # Cancel any tasks that are slicing a subset of the layers
+            # being sliced now. This allows us to slice arbitrary sets of
+            # layers with some sensible and not too complex cancellation
+            # policy.
+            layer_set = set(layers)
+            for task_layers, task in self._layers_to_task.items():
+                if set(task_layers).issubset(layer_set):
+                    LOGGER.debug('Cancelling task for %s', task_layers)
+                    task.cancel()
+
+            # Not all layer types will initially be asynchronously sliceable.
+            # The following logic gives us a way to handle those in the short
+            # term as we develop, and also in the long term if there are cases
+            # when we want to perform sync slicing anyway.
+            requests = {}
+            for layer in layers:
+                # TODO: temmporary to prove a point
+                # assert isinstance(layer, weakref.ReferenceType) # this fails
+                if layer._is_async():
+                    requests[layer] = layer._make_slice_request(dims)
+                else:
+                    layer._slice_dims(dims.point, dims.ndisplay, dims.order)
+            # create task for slicing of each request/layer
+            task = self._executor.submit(self._slice_layers, requests)
```

Comment:
@andy-sweet won't this submit the slicing of all layers under one thread?

Shouldn't it be something like:
`future_dict = {self._executor.submit(self._slice_layers, req): req for req in requests]` ? In order to submit them to separate threads?

Reply:
Actually, more like this:
`{self._executor.submit(self._slice_layers, req): layer for req, layer in requests.items()}`

Reply:
that actually turns out messy since I still need to `add_done_callback`. This seems to work better 

```
        for request, layer in requests.items():
            future = self._executor.submit(self._slice_layers, request)
            future.add_done_callback(self._on_slice_done)
            self._futures[future] = layer
```
I'm toying with the idea of replacing `self.task_to_layers` with `self._futures`. `task_to_layers` comes with a lot of weird baggage:
* what if the next slice is on a different group of layers, we won't match on the keys?
* the minimal lock I added based on your suggestion has some flaws and I suspect its not sufficient, we'd need to cover almost all the code in a lock like I had before
* layer tuples are not unique, but futures are

Reply:
> @andy-sweet won't this submit the slicing of all layers under one thread?

Yes. There's only one slicing thread because we initialize `ThreadPoolExecutor` with `max_workers=1`.

> Shouldn't it be something like:
> `future_dict = {self._executor.submit(self._slice_layers, req): req for req in requests]` ? In order to submit them to separate threads?

That will create a future/task/job for each layer. Initially, I avoided that to avoid previously reported jarring visualization issues (see #2862). These can be particularly bad when using the slider to navigate through 2D slices of 3D data, which is why changing `current_step` slices all the layers together in one thread.

I agree that if we don't care about that behavior, we can simplify the implementation here, and also have fine grained cancellation. Though if we don't care about that, I think that [the suggested vispy driven slicing](https://napari.org/dev/naps/4-async-slicing.html#slice-from-vispy-layer-instead-of-viewer-model) might be the way to go.

Note that there is a related [discussion in the NAP around rendering slices as soon as they are ready](https://napari.org/dev/naps/4-async-slicing.html#render-each-slice-as-soon-as-it-is-ready), though it is a little different.

> * what if the next slice is on a different group of layers, we won't match on the keys and the pending slices wont get cancelled.

Mostly seems fine to me. As you've seen, cancelling only works for pending tasks, unless we do something extra ourselves. Having more fine-grained, eager cancellation could be nice, but it may also complicate things and result in a less interactive experience (which is what I found qualitatively).

> * the minimal lock I added based on your suggestion has some flaws and I suspect its not sufficient, we'd need to cover almost all the code in a lock like I had before

Can you explain why? Ideally with a failing test. My feeling is that we probably don't need to hold that lock over too much code, but I might be wrong. If we need to, that should be OK because `slice_layers_async` should finish quickly.

> * layer tuples are not unique, but futures are

That makes it trivial to remove the future in `_on_slice_done`, but harder to deal with in `slice_layers_async` right? The right choice here also depends on what mapping we care about (i.e. future <-> tuple[layer] or future <-> layer).